### PR TITLE
ci: Added Spell checker in CI

### PR DESCRIPTION
**Pull request by aakash-pamnani**
# Description
Added spell check in Github CI

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

## Related Issues
#1809 

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Comment by st-pasha** [comment]
So, there are now lots of errors reported by the spellchecker:
1. Real typos such as "contollable" -- these need to be fixed;
2. Words that are not known to the spellchecker but make sense in their context, such as "abcdefghijklmnopqrstuvwxyz" -- these can be added as exceptions;
3. Words that shouldn't have been spell-checked in the first place, such as URL fragments "positioncomponent". These ones I'm not sure how to deal with - is there a way to instruct the spell checker not to include URLs in the check?

**Comment by aakash-pamnani** [comment]
> So, there are now lots of errors reported by the spellchecker:
> 
> 1. Real typos such as "contollable" -- these need to be fixed;
> 2. Words that are not known to the spellchecker but make sense in their context, such as "abcdefghijklmnopqrstuvwxyz" -- these can be added as exceptions;
> 3. Words that shouldn't have been spell-checked in the first place, such as URL fragments "positioncomponent". These ones I'm not sure how to deal with - is there a way to instruct the spell checker not to include URLs in the check?

2 - > these words can be added to ignoredWords list in cspell.json
3-> we can use a field called regexignore to ignore words with regEx match

**Comment by aakash-pamnani** [comment]
> So, there are now lots of errors reported by the spellchecker:
> 
> 1. Real typos such as "contollable" -- these need to be fixed;
> 2. Words that are not known to the spellchecker but make sense in their context, such as "abcdefghijklmnopqrstuvwxyz" -- these can be added as exceptions;
> 3. Words that shouldn't have been spell-checked in the first place, such as URL fragments "positioncomponent". These ones I'm not sure how to deal with - is there a way to instruct the spell checker not to include URLs in the check?

https://github.com/streetsidesoftware/cspell/tree/main/packages/cspell#exclude-patterns
 according to this urls are already excluded by default

**Comment by st-pasha** [comment]
> according to this urls are already excluded by default

But it probably doesn't know about markdown links, which look like this: `[](#spriteanimationcomponent)`.

On the other hand, I see the `enableCompoundWords` option - this might solve the problem (though we'd need to enable it for md files only).

File: .github/cspell.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+    "version": "0.1",
+    "language": "en",
+    "flagWords": [],
+    "ignoreWords":["broadphase","subdir","Hitbox","Raycast","positionable","Renderable","pallete","raytracing"] ,
```
Comment (st-pasha) [inline .github/cspell.json:5]: 1. `pallete` is a typo - should be "palette". Also, `subdir` is just an abbreviation of "subdirectory".
2. This looks like it'll end up being a rather long list, so let's make it one-word-per-line to make future edits easier.
3. Is there a difference between capitalized and lowercase words? If not, then probably all words should be turned into lowercase.

Comment (aakash-pamnani) [reply .github/cspell.json:5]: 1-> just added few words from the output of the spell checker
2-> sure

3-> it is case insensitive (https://github.com/streetsidesoftware/cspell/tree/main/packages/cspell#things-to-note)

Comment (spydon) [reply .github/cspell.json:5]: What is left to do here?

Comment (st-pasha) [reply .github/cspell.json:5]: There are 2526 issues found by the analyzer :)

Comment (aakash-pamnani) [reply .github/cspell.json:5]: What should we do to solve this?
Adding every word in ignore words  list?

Comment (spydon) [reply .github/cspell.json:5]: I guess so, but we'll have to make sure to not add actual typos to the ignore list so we can't just do a blanket add.

Comment (aakash-pamnani) [reply .github/cspell.json:5]: We can use the spell checker to check our ignore words list then 😂

Comment (st-pasha) [reply .github/cspell.json:5]: I can see plenty of typos reported: `programatically`, `operatiosn`, `paramater`, `slided`, `everytime`, `Uncontolled`, etc. -- that's just among the last 20.

Comment (aakash-pamnani) [reply .github/cspell.json:5]: I will try to add as much i can in the list and also correct them in the files.

**Comment by aakash-pamnani** [comment]
Is it ok to ignore all this files?
```  
"ignorePaths": ["./media/**","./**/*.ogg","./**/*.svg","./**/*.mp3","./**/*.json","./**/license.txt","*/assets/"],
```
and also can we ignore test folder?

**Comment by aakash-pamnani** [comment]
> Is it ok to ignore all this files?
> 
> ```
> "ignorePaths": ["./media/**","./**/*.ogg","./**/*.svg","./**/*.mp3","./**/*.json","./**/license.txt","*/assets/"],
> ```
> 
> and also can we ignore test folder?

do we only need to scan .md and .dart files?

**Comment by munsterlander** [comment]
I'm somewhat worried about this. Grammarly goes nuts with all the individual nuances of Flame.  Obviously component names, but even in back ticked words like `fileName.dart` get flagged. This is going to be interesting. I do think you only need to scan dart and markdown files though.

File: .github/cspell.json [hunk 2]
```diff
@@ -1,8 +1,37 @@
...
+    "pressable",
+    "stylesheet",
+    "Strikethrough",
+    "blockquote",
+    "proxying",
+    "flippable",
+    "ianliu",
+    "renancarujo",
+    "illiapoplawski",
+    "audiopool"
```
Comment (st-pasha) [inline .github/cspell.json:28]: Can be replaced with `AudioPool` I think.

**Comment by st-pasha** [comment]
@spydon I wonder if it's possible to add this `spell-checker` action as-is (without fixing all the spell errors), but mark it as non-blocking for now. That is, if the check fails, as it does right now, the PR would still be considered "green".

I'm worried that fixing all these spell errors is a bit more complicated than we anticipated (over 2700 problems reported), so it might be better to do that over a series of PRs.

File: examples/games/padracing/lib/padracing_game.dart [hunk 3]
```diff
@@ -175,7 +175,7 @@ class PadRacingGame extends Forge2DGame with KeyboardEvents {
         if (car.lapNotifier.value > numberOfLaps) {
           isGameOver = true;
           winner = car;
-          overlays.add('gameover');
+          overlays.add('game over');
```
Comment (spydon) [inline examples/games/padracing/lib/padracing_game.dart:178]:
```suggestion
          overlays.add('gameover');
```
Reverting this since it is breaking

**Comment by spydon** [comment]
> @spydon I wonder if it's possible to add this `spell-checker` action as-is (without fixing all the spell errors), but mark it as non-blocking for now. That is, if the check fails, as it does right now, the PR would still be considered "green".
> 
> I'm worried that fixing all these spell errors is a bit more complicated than we anticipated (over 2700 problems reported), so it might be better to do that over a series of PRs.

Yeah, that makes sense.
Reverted the game over layover change since that was breaking.

**Review by spydon — APPROVED** [review]
Good job @aakash-pamnani!
If you can just fix the open comments from Pasha I'll merge this in and we'll open another issue for going through all the rest.

File: packages/flame/lib/src/parallax.dart [hunk 4]
```diff
@@ -195,13 +195,13 @@ class ParallaxAnimation extends ParallaxRenderer {
 
     final animation =
         await SpriteAnimation.load(path, animationData, images: images);
-    final prerendedFrames = await Future.wait(
+    final preRendedFrames = await Future.wait(
```
Comment (spydon) [inline packages/flame/lib/src/parallax.dart:198]: This should be the same as in the other parallax class

File: tutorials/space_shooter/lib/steps/1_getting_started/2_input_and_graphics/tutorial.dart [hunk 5]
```diff
@@ -139,7 +139,7 @@ Just like `FlameGame`, components also have an `onLoad` method that can be overr
 initializations. But before we implement our player's own load method, note that we use an attribute
 and the `loadSprite` method from the `FlameGame` class.
 
-That is not a problem! Everytime our component needs to access things from its game class, we can
+That is not a problem! every time our component needs to access things from its game class, we can
```
Comment (spydon) [inline tutorials/space_shooter/lib/steps/1_getting_started/2_input_and_graphics/tutorial.dart:142]: Should be capitalized

File: packages/flame/lib/src/parallax.dart (hunk 4 above)
Comment (aakash-pamnani) [reply packages/flame/lib/src/parallax.dart:198]: sorry, forget that 
will fix that

File: tutorials/space_shooter/lib/steps/1_getting_started/2_input_and_graphics/tutorial.dart (hunk 5 above)
Comment (aakash-pamnani) [reply tutorials/space_shooter/lib/steps/1_getting_started/2_input_and_graphics/tutorial.dart:142]: E should be capitalized?

Comment (spydon) [reply tutorials/space_shooter/lib/steps/1_getting_started/2_input_and_graphics/tutorial.dart:142]: Yes

File: CHANGELOG.md [hunk 6]
```diff
@@ -39,7 +39,7 @@ Packages with dependency updates only:
 
 #### `flame` - `v1.4.0`
 
- - **REFACTOR**: move broadphase-related functionality into separate subdir ([#1943](https://github.com/flame-engine/flame/issues/1943)). ([f23acd41](https://github.com/flame-engine/flame/commit/f23acd41e0341909200437bfb6487cbe9ca58a53))
+ - **REFACTOR**: move broadphase-related functionality into separate subdirectory ([#1943](https://github.com/flame-engine/flame/issues/1943)). ([f23acd41](https://github.com/flame-engine/flame/commit/f23acd41e0341909200437bfb6487cbe9ca58a53))
```
Comment (spydon) [inline CHANGELOG.md:42]: Why does it report on all these words that are already in the ignore list? :thinking:

Comment (spydon) [reply CHANGELOG.md:42]: image

Comment (aakash-pamnani) [reply CHANGELOG.md:42]: this does not happens when i run cspell locally
may be there is some issue with the configuration file path

File: packages/flame/CHANGELOG.md [hunk 7]
```diff
@@ -983,7 +983,7 @@
 - Svg support
 - Adding `Animation#reversed` allowing a new reversed animation to be created from an existing animation.
 - Fix games inside regular apps when the component is inside a sliver
-- Support asesprite animations
+- Support aseprite animations
```
Comment (st-pasha) [inline packages/flame/CHANGELOG.md:986]: I believe Aseprite should be capitalized here.

File: .github/cspell.json [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    "Strikethrough",
+    "stylesheet",
+    "vsync"
+  ],
+  "ignorePaths": [
+    "**/media/**",
+    "**/assets/**",
+    "**/test/**"
+  ],
+  "ignoreRegExpList":["@\\w+"],
```
Comment (st-pasha) [inline .github/cspell.json:30]: nit: space after `:`

**Review by st-pasha — APPROVED** [review]

File: .github/cspell.json [hunk 9]
```diff
@@ -0,0 +1,32 @@
+{
+  "version": "0.1",
+  "language": "en",
+  "flagWords": [],
+  "ignoreWords": [
```
Comment (st-pasha) [inline .github/cspell.json:5]: Looking at https://cspell.org/configuration/, I believe this parameter should be called just `words`. Also, the `version` should be 0.2.

**Comment by aakash-pamnani** [comment]
after changing version and ```ignoreWords``` to ```words``` its still not working

File: .github/workflows/spell_checker.yml [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: zwaldowski/cspell-action@v1
+        with:
+          paths: "**/*.{md,dart}"
+          config: .github/cspell.json /// This file would include spelling exceptions
```
Comment (spydon) [inline .github/workflows/spell_checker.yml:17]:
```suggestion
          config: .github/cspell.json
```

File: .github/workflows/spell_checker.yml [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: zwaldowski/cspell-action@v1
+        with:
+          paths: "**/*.{md,dart}"
+          config: .github/cspell.json
+          exclude: ".gitignore"
```
Comment (spydon) [inline .github/workflows/spell_checker.yml:18]:
```suggestion
          exclude: ".gitignore,CHANGELOG.md"
```

File: .github/cspell.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  "words": [
+    "ASDW",
+    "blockquote",
+    "broadphase",
+    "dashbook",
+    "flippable",
+    "hitbox",
+    "LTRBR",
+    "LTWH",
+    "positionable",
```
Comment (spydon) [inline .github/cspell.json:14]:
```suggestion
    "markdownlint",
    "positionable",
```

File: .github/cspell.json [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    "proxying",
+    "raycast",
+    "raytracing",
+    "renderable",
+    "Strikethrough",
+    "stylesheet",
+    "vsync"
+  ],
+  "ignorePaths": ["**/media/**", "**/assets/**", "**/test/**"],
+  "ignoreRegExpList": ["@\\w+"],
```
Comment (spydon) [inline .github/cspell.json:27]:
```suggestion
  "ignoreRegExpList": ["@\\w+", "\\$\\w+", "\\{\\w+", "\\`\\w+"],
```

**Comment by aakash-pamnani** [comment]
Was it not working due to the comment in the cspell.json file?

**Comment by spydon** [comment]
> Was it not working due to the comment in the cspell.json file?

Yes exactly, json doesn't have a concept of comments.
The only thing left for this one is to make the action mark it green even though it is failing.

**Review by erickzanardo — APPROVED** [review]

**Pull request merged**